### ISSUE TITLE
setupController doesn't fire after model refresh

### ISSUE THREAD

**Issue by dmagunov** [emberjs/ember.js#11442]
Ember 1.12.1 when updating route model on the same route (for example via queries) setupController doesn't call.

```coffeescript
model: ->
	return Ember.$.getJSON("http://some.url").then (json)=>
		obj = {}
		for name, val of json
			@store.pushMany  name.camelize(), val
			obj[name.camelize()] = @store.all(name.camelize()) 
		@set('dict', obj)
		return @store.all('project')

setupController: (controller, model)->
		
	@_super(controller, model)
	controller.set('dict', @get('dict'))
```

**Comment by jslippy** [comment]
Is anybody working on this?  It is still an issue in Ember 1.13.5.  When I call `refresh()` the `model` hook fires but `setupController` doesn't fire.

**Comment by jcope2013** [comment]
@jslippy setupController only runs once on controller initialization, refresh() only triggers the model hooks

**Comment by fedyaFade** [comment]
http://emberjs.com/api/classes/Ember.Route.html#method_refresh
>setupController and associated route hooks will re-fire as well

**Comment by stefanpenner** [comment]
I if the model hook is called setup controller should also be called. If an example can be provided that demonstrates the issue, that would be a good place to start debugging

**Comment by PoplarK** [comment]
I get the same problem. When I use 'return  {type: "refresh"}' in 'model' function to replace a promise object like 'return $.ajax(xxxxx)' , 'setupController' can be called. But I don't know the reason. Here is my code of 'model' function:
model: function(params) { return {type: "refresh"} };

**Comment by amedrz** [comment]
@jslippy @PoplarK we had a similar problem with this code which wasn't triggering `setupController`:

````
model() {
  return this.store.findRecord('account', 'me')
  .then(
    (account) => {
      this.set('currentAccount', account);
      this.mixpanel.identify(account);
    }
  );
}
````
`this.mixpanel.identify(account);` is an injected service which would do their job and return `undefined` which would cause the promise don't go along and thus `setupController` wouldn't be called.

In promises, you better return either another promise or a synchronous value, so I just switched the order of the functions inside `then()` and it worked.

````
model() {
  return this.store.findRecord('account', 'me')
  .then(
    (account) => {
      this.set('currentAccount', account);
      this.mixpanel.identify(account);
    }
  );
}
````
For more info: http://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html

**Comment by laureanoarcanio** [comment]
Any updates / workaround about this ? It's seems to be present on Ember 2.3.0.

**Comment by locks** [comment]
@laureanoarcanio can you provide an Ember Twiddle reproducing the problem?

**Comment by pixelhandler** [comment]
@dmagunov do you still have this issue? If so can you provide an example reproduction, e.g. ember-twiddle or other online tools, or even a simple git repo?

**Comment by dmagunov** [comment]
No, I don't (in 2.6)

**Comment by duart** [comment]
Can be  off-topic but I landed here with the same symptoms.
Check if it is not the problem mentioned in [https://just4fun.github.io/hexo-blog/2017/01/21/13-reloading-in-ember-findall-and-findrecord/](url).
I faced situation where model seems to not be reloaded. I though that setupControler was not fired as I did not see changed data in conttroller, but what was strange, I could see changed data in Ember Inspector. 
What really happened is that findAll promise in model event were resolved immediatelly from existing store data (so setupController was fired with old data) and in meantime store was reloading them from api. As an effect I had to call refresh twice to get data available in controller.
The solution was to add " { reload: true }" to findAll params.
